### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by gogonzo**
Part of https://github.com/insightsengineering/teal.data/issues/102
Other PRs https://github.com/insightsengineering/teal/pull/899 https://github.com/insightsengineering/teal.slice/pull/440

## install
<details>
<summary>Installation code </summary>

```r
staged.dependencies::dependency_table(
  project = "insightsengineering/teal@https://github.com",
  project_type = "repo@host",
  ref = "refactor",
  verbose = 1
) |> staged.dependencies::install_deps()
```
</details>

## ddl class

<details><summary>unfold</summary>

Small class which contains `ui` and `server`. `server` calls `run_ddl` which evaluates a `code` using input and create `teal_data` object. Function `ddl` accepts several arguments which are used to generate proper `ui` and `server`.

tdata-DDL process (3)

Above diagram depict how class works. Please follow steps below and look at the diagram. 

1. One needs to specify `ddl` arguments to fit someone's needs. Take for example `custom_connector`. Let's assume that we need to call `scda::synthetic_cdisc_data('latest')` for two datasets.

```r
data_code = paste(
    "adsl <- scda::synthetic_cdisc_data('latest')$adsl",
    "adtte<- scda::synthetic_cdisc_data('version')$adtte",
    collapse = "\n"
  )
)
```

Imagine that above datasets are on the remote server so we need to call `open_conn` function and close after pull. Imagine that `open_conn` and `close_conn` are in some external package. 

```r
code = paste(c(
  "open_conn(username = input$username, password = input$password)",
  "adsl <- scda::synthetic_cdisc_data('latest')$adsl",
  "adtte<- scda::synthetic_cdisc_data('version')$adtte",
  "close_conn()"
  ),
  collapse = "\n"
)
```

`username` and `password` should be specified by the app user. To do so we need to create `ui` which will create `username` and `password` inputs which are used during evaluation of above code.

```r
ui = function(id) {
  ns <- NS(id)
  tagList(
    textInput(ns("username"), label = "Username"),
      passwordInput(ns("password"), label = "Password"),
      actionButton(ns("submit"), label = "Submit")
  )
}
```

If we have `ui` we need also `server` to utilize input. When submit button is clicked all `input` values are passed to `ddl_run` where all magic happens. `ddl_run` executes the code and substitutes all `input$<name>` with equivalent `input_list` element. 

```r
submit_button_server <- function(id) {
  moduleServer(id, function(input, output, session) {
    tdata <- eventReactive(input$submit, {
      req(input$pass)
      ddl_run(input_list = reactiveValuesToList(input))
    })

    # would need to make sure we handle reactivity correctly here as teal::init expects not reactive tdata...
    return(tdata)
  })
}
```

Because user `username` and `password` is not something we would like to include in the show R code. We need to replace them with some other code which can be run by anyone who receives the code. To replace args with some desired value or call we use `offline_args` to substitute code with these.

```
offline_args = list(
  username = quote(askpass::askpass("Please enter username")),
  password = quote(askpass::askpass("Please enter password"))
)  
```

At this point `ddl_run` evaluated "online" code, and has environment with the results of the evaluation. Also "offline" code is available after substitution. At the end `teal_data_new` is constructed from data and offline code

```r
teal_data_new(env_list, code = masked_code, join_keys)
```

⚠️ As one can see `ddl_run` require only `input_list`. It is a simplification for potential app developer who doesn't need to handle additional arguments (`code`, `input_mask`) in the `server`. Consequence of this simplification is that `server` specified in `ddl()` have an enclosing environment changed to the environment where `ddl_run`, `code`, `input_mask` is available. Another consequence is that `ddl_run` theoretically doesn't exist anywhere - it is created on the fly when `ddl` object is created. It means that developing own `ddl` connector in external "pkg" can't use private functions from that "pkg".

</details>

## Example connector app

<details>
<summary>R code</summary>

```r
custom_password_connector <- function(datanames) {
  ddl(
    # code to be run when app user presses submit
    code = paste(
      c(
        "conn <- open_conn(username = input$username, password = input$password)",
        sprintf(
          "%1$s <- scda::synthetic_cdisc_data('latest')$%1$s",
          datanames
        ),
        "close_conn()"
      ),
      collapse = "\n"
    ),

    # arguments used for show R code
    mask_args = list(
      username = quote(askpass::askpass()),
      password = quote(askpass::askpass())
    ),

    # ui they wish to use for the loading data
    ui = function(id) {
      ns <- NS(id)
      tagList(
        textInput(ns("username"), label = "Username"),
        passwordInput(ns("password"), label = "Password"),
        actionButton(ns("submit"), label = "Submit")
      )
    },
    datanames = datanames
  )
}

pkgload::load_all("teal.data")
pkgload::load_all("teal")
library(shiny)

x <- custom_password_connector(c("adsl", "adtte", "adrs"))


app <- shinyApp(
  ui = fluidPage(
    fluidRow(
      column(3, h1("User Inputs"), x$ui(id = "custom_ui")),
      column(9, h1("R code"), verbatimTextOutput("output"))
    )
  ),
  server = function(input, output, session) {
    loaded_data <- x$server(id = "custom_ui", x = x)
    output$output <- renderText({
      req(loaded_data())
      teal.code::get_code(loaded_data()) |> paste(collapse = "\n")
    })
  }
)

shiny::runApp(app)
```

</details>

## Example teal app with connector

<details>
<summary>teal app code</summary>

use `teal.data`, `teal` and `teal.slice` from `refactor` branch

```r
# options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
options("teal.bs_theme" = bslib::bs_theme(version = 3))

library(shiny)
library(scda)
library(scda.2022)
library(teal.data)
library(teal.transform)
pkgload::load_all("teal.data")
pkgload::load_all("teal.slice")
pkgload::load_all("teal")

library(teal.modules.general)

funny_module <- function(label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    datanames = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <- renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

custom_password_connector <- function(datanames) {
  ddl(
    # code to be run when app user presses submit
    code = paste(
      c(
        "conn <- open_conn(username = input$username, password = input$password)",
        sprintf(
          "%1$s <- scda::synthetic_cdisc_data('latest')$%1$s",
          datanames
        ),
        "close_conn()"
      ),
      collapse = "\n"
    ),

    # arguments used for show R code
    mask_args = list(
      username = quote(askpass::askpass()),
      password = quote(askpass::askpass())
    ),

    # ui they wish to use for the loading data
    ui = function(id) {
      ns <- NS(id)
      tagList(
        textInput(ns("username"), label = "Username"),
        passwordInput(ns("password"), label = "Password"),
        actionButton(ns("submit"), label = "Submit")
      )
    },
    datanames = datanames
  )
}
data <- custom_password_connector(c("adsl", "adtte", "adrs"))

app <- init(
  data = data,
  modules = modules(
    funny_module("funny"),
    funny_module("funny2", datanames = "adtte") # will limit datanames to ADTTE and ADSL (parent)
  )
)

runApp(app)
```

</details>

## teal app with teal_data

<details>
<summary>App code</summary>

```r
# options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
options("teal.bs_theme" = bslib::bs_theme(version = 3))

library(shiny)
library(scda)
library(scda.2022)
library(teal.data)
library(teal.transform)
pkgload::load_all("teal.data")
pkgload::load_all("teal.slice")
pkgload::load_all("teal")
library(teal.modules.general)

funny_module <- function(label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    datanames = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <- renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

data <- cdisc_data(
  ADSL = synthetic_cdisc_data("latest")$adsl,
  ADTTE = synthetic_cdisc_data("latest")$adtte,
  ADRS = synthetic_cdisc_data("latest")$adrs,
  code = '
    ADSL <- synthetic_cdisc_data("latest")$adsl
    ADTTE <- synthetic_cdisc_data("latest")$adtte
    ADRS <- synthetic_cdisc_data("latest")$adrs
  '
)

app <- init(
  data = data,
  modules = modules(
    funny_module("funny"),
    funny_module("funny2", datanames = "ADTTE") # will limit datanames to ADTTE and ADSL (parent)
  )
)

runApp(app)

```

</details>

## teal.gallery examples

- exploratory app with tdata https://github.com/insightsengineering/teal.gallery/blob/refactor/exploratory/app.R
- basic teal app with ddl https://github.com/insightsengineering/teal.gallery/blob/refactor/basic-teal/app.R

## Internal Connectors

https://github.com/insightsengineering/nestdevs-tasks/issues/38

## What's next

To complete #102 we need to do following:

Match arguments across all reproducible objects/classes (ddl, qenv, teal_data):

- https://github.com/insightsengineering/teal.code/issues/145
- https://github.com/insightsengineering/teal.data/issues/169
- https://github.com/insightsengineering/teal.code/issues/127

Other:

- https://github.com/insightsengineering/teal/issues/904
- https://github.com/insightsengineering/teal.slice/issues/132
- https://github.com/insightsengineering/teal.transform/issues/111
- https://github.com/insightsengineering/teal/issues/669
- https://github.com/insightsengineering/teal.data/issues/162
- https://github.com/insightsengineering/teal/issues/860 
- (possibly) https://github.com/insightsengineering/teal/issues/534

### CODE REVIEW

**Comment by m7pr** [comment]
image

**Comment by gogonzo** [comment]
@m7pr 
> I'm getting below, when I run Example connector app or Example teal app with connector

Yes, you typed any password. Message says that password is `pass` ;)

image

**Comment by m7pr** [comment]
So it looks like it did the job, but it thrown some other unrelated error that confused me?

**Comment by averissimo** [comment]
@m7pr I needed to add `pkgload::load_all("teal.slice")` for it to work flawlessly

File: R/teal_data.R [hunk 1]
```diff
@@ -36,27 +36,41 @@ teal_data <- function(...,
...
+      new_tdata(
+        env = lapply(x$get_datasets(), function(x) x$get_raw_data()),
+        code = x$get_code(),
+        keys = x$get_join_keys()
+      )
+    } else {
+      x
+    }
+  } else {
+    new_tdata(
```
Comment (averissimo) [inline R/teal_data.R:65]: What do you think about protecting against unnamed data parameters?

The error is not very user friendly

> names(x) must be a character vector of the same length as x

2 suggestions below if this makes sense:

### 1. outputs an error if arguments are not named


```suggestion
    new_tdata(
    if (!checkmate::test_names(names(data_objects), type = "named")) {
      stop("Dot (`...`) arguments on `teal_data()` must be named.")
    }
```

### 2. tries to extrapolate variable names and use it as name

```R
new_iris <- transform(iris, id = seq_len(nrow(iris)))
new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))

data = teal_data(
  new_iris,
  new_mtcars,
  code = "
      new_iris <- transform(iris, id = seq_len(nrow(iris)))
      new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))
    "
)

data = teal_data(
  new_iris = new_iris,
  new_mtcars,
  code = "
      new_iris <- transform(iris, id = seq_len(nrow(iris)))
      new_mtcars <- transform(mtcars, id = seq_len(nrow(mtcars)))
    "
)
```

```suggestion
    new_tdata(
    if (!checkmate::test_names(names(data_objects), type = "named")) {
      tryCatch({
        var_names <- vapply(substitute(list(...))[-1], deparse, character(1))
        unamed_ix <- names(data_objects) == ""
        names(data_objects)[unamed_ix] <- var_names[unamed_ix]
      }, error = function(err) stop("Dot (`...`) arguments on `teal_data()` must be named.")
      )
    }
```

**Comment by m7pr** [comment]
Still no luck getting this properly build. I just run staged.dependencies builds

```{R}
x <- 
  staged.dependencies::dependency_table(
    project = "insightsengineering/teal@https://github.com",
    project_type = "repo@host",
    ref = "refactor",
    verbose = 1
  )
x |> staged.dependencies::install_deps()

x <- 
  staged.dependencies::dependency_table(
    project = "insightsengineering/teal.data@https://github.com",
    project_type = "repo@host",
    ref = "refactor",
    verbose = 1
  )

x |> staged.dependencies::install_deps()
```

Run the app 

```r
# options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
options("teal.bs_theme" = bslib::bs_theme(version = 3))

library(shiny)
library(scda)
library(scda.2022)
library(teal.data)
library(teal.transform)
pkgload::load_all(".")
pkgload::load_all("../teal")
library(teal.modules.general)

funny_module <- function(label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    datanames = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <- renderPrint({
        logger::log_trace("rendering text1")
        filter_panel_api %>% get_filter_state()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

data <- cdisc_data(
  ADSL = synthetic_cdisc_data("latest")$adsl,
  ADTTE = synthetic_cdisc_data("latest")$adtte,
  ADRS = synthetic_cdisc_data("latest")$adrs,
  code = '
    ADSL <- synthetic_cdisc_data("latest")$adsl
    ADTTE <- synthetic_cdisc_data("latest")$adtte
    ADRS <- synthetic_cdisc_data("latest")$adrs
  '
)

app <- init(
  data = data,
  modules = modules(
    funny_module("funny"),
    funny_module("funny2", datanames = "ADTTE") # will limit datanames to ADTTE and ADSL (parent)
  )
)

runApp(app)

```

and got the following

```r
Error in cdisc_data(ADSL = synthetic_cdisc_data("latest")$adsl, ADTTE = synthetic_cdisc_data("latest")$adtte, :
Assertion on 'data_objects' failed: May only contain the following types: {TealDataset,TealDatasetConnector,TealDataConnector}, but element 1 has type 'tbl_df,tbl,data.frame'.
```

**Comment by m7pr** [comment]
I'm getting something weird in here

```r
data <- cdisc_data(
  ADSL = synthetic_cdisc_data("latest")$adsl,
  ADTTE = synthetic_cdisc_data("latest")$adtte,
  ADRS = synthetic_cdisc_data("latest")$adrs,
  code = '
    ADSL <- synthetic_cdisc_data("latest")$adsl
    ADTTE <- synthetic_cdisc_data("latest")$adtte
    ADRS <- synthetic_cdisc_data("latest")$adrs
  '
)
```

**Comment by m7pr** [comment]
I did checkout to `refactor` branch locally for `teal`, `teal.slice` and `teal.data` made pulls and build those packages locally. I thought `staged.dependencies` would do that for me. I actually think that `pkgload::load_all('teal')` was loading from improper branch on my local machine.

I was able to run the `Example connector app`
image

and `Example teal app with connector`
image

File: R/cdisc_data.R [hunk 2]
```diff
@@ -68,58 +68,80 @@ cdisc_data <- function(...,
...
+        code = x$get_code(),
+        keys = x$get_join_keys()
+      )
+    } else {
+      x
+    }
+  } else {
+    if (!checkmate::test_names(names(data_objects), type = "named")) {
+      stop("Dot (`...`) arguments on `teal_data()` must be named.")
+    }
```
Comment (chlebowa) [inline R/cdisc_data.R:142]:
```suggestion
    checkmate::assert_names(names(data_objects), .var.name = "dot arguments (`...`) in `teal_data()`")
```

Comment (averissimo) [reply R/cdisc_data.R:142]: Ah great!

The function I used before is the soon-to-be-deprecated `assert_named()` which had a non-descriptive error message (`failed: Must have Object.`), but this is much cleaner :-)

**Review by vedhav — COMMENTED** [review]
I know it's still in the draft and looks like there do not seem to be major changes to the PR.

Can we update the `_pkgdown.yml` with the new exported functions and add some vignettes to explain the new functions?

File: R/ddl.R [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  }
+
+  # don't pass non-dataset bindings further
+  # we don't want to initialize tdata with them
+  env_list <- env_list[x$datanames]
+
+  # substitute by offline args
+  for (i in names(x$offline_args)) {
+    online_args[[i]] <- x$offline_args[[i]]
+  }
```
[the remark this answers is not in the document]
Comment (m7pr) [reply R/ddl.R:102]: Great simplification @chlebowa . This works even for NULL (empty) `offline_args`

```r
online_args <- list(password = 'asd', user = 'mk')
x <- list(offline_args = NULL)
online_args2 <- online_args

for (i in names(x$offline_args)) {
  online_args[[i]] <- x$offline_args[[i]]
}


online_args2[names(x$offline_args)] <- x$offline_args

identical(online_args2, online_args)
```

File: R/tdata.R [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    id <- sample.int(.Machine$integer.max, size = length(code))
+    methods::new(
+      "tdata",
+      env = new_env,
+      code = code,
+      warnings = rep("", length(code)),
+      messages = rep("", length(code)),
+      id = id,
+      join_keys = keys,
+      datanames = union(names(env), names(keys$get()))
```
Comment (chlebowa) [inline R/tdata.R:50]:
```suggestion
      datanames = as.character(union(names(env), names(keys$get())))
```

File: R/tdata.R [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#' Initialize `tdata` object.
+#' @name new_tdata
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Accepts and stores comments also.
+#' @param env (`list`) List of data.
+#'
+#' @examples
+#' new_tdata(env = list(a = 1), code = quote(a <- 1))
+#' new_tdata(env = list(a = 1), code = parse(text = "a <- 1"))
+#' new_tdatas(env = list(a = 1), code = "a <- 1")
```
Comment (chlebowa) [inline R/tdata.R:26]:
```suggestion
#' new_tdata(env = list(a = 1), code = "a <- 1")
```

File: R/cdisc_data.R [hunk 6]
```diff
@@ -68,58 +68,80 @@ cdisc_data <- function(...,
...
+      "cdisc_data(
+        data_objects = 'should use data directly. Using TealDatasetConnector and TealDataset is deprecated.'
+      )"
     )
-  }
+      tryCatch(
+        x$get_parent(),
+        error = function(cond) rep(character(0), length(x$get_datanames()))
+      )
+    }
```
Comment (chlebowa) [inline R/cdisc_data.R:93]: Seems like the `tryCatch` could be put directly into the anonymous function in `lapply` in `new_parents_fun`.

File: R/is_pulled.R
Comment (chlebowa) [inline R/is_pulled.R:1]: This function will go, right?

**Comment by gogonzo** [comment]
Closing this PR in favour of  https://github.com/insightsengineering/teal.data/pull/171

**Pull request closed without merging**